### ISSUE TITLE
Hackney dies after too many 404 responses

### ISSUE THREAD

**Issue by jaseemabid** [benoitc/hackney#284]
The following line of code never returns.

`[hackney:get("http://google.com/i-am-a-404") || _ <- lists:seq(1,50)].`

For comparison, apache bench completes the task in about 9s on the same machine.

```
$ ab -n 50 http://google.com/i-am-a-404
Time taken for tests:   9.603 seconds
Complete requests:      50
Failed requests:        0
Non-2xx responses:      50
Requests per second:    5.21 [#/sec] (mean)
Time per request:       192.058 [ms] (mean)

Connection Times (ms)
              min  mean[+/-sd] median   max
Connect:       23   29   5.4     27      49
Processing:   155  163   5.5    161     187
Waiting:      154  162   5.6    160     186
Total:        181  192   7.5    190     218

Percentage of the requests served within a certain time (ms)
  50%    190
  66%    193
  75%    196
  80%    198
  90%    201
  95%    206
  98%    218
  99%    218
 100%    218 (longest request)
```

**Comment by swvist** [comment]
Update:

This line doesn't wait indefinitely:

```erlang
[hackney:get("http://google.com/i-am-a-404") || _ <- lists:seq(1,50)].
%% 50 requests
```

This does:
```erlang
[hackney:get("http://google.com/i-am-a-404") || _ <- lists:seq(1,51)].
%% 51 requests
```

This happens because hackney pool is by default on with `max_connections` set to 50. A socket is returned to the pool only when a request is made *and* the body is consumed. Since, this a two step process, and for someone who is only interested in the status code, it is not immediately obvious that one *has* to call `hackney:body/1` regardless.

We can do either of these:
1) Document that calling `hackney:body/1` is required.
2) Add a request option can be added to hackney, if its not already present, that can ignore the body when you are only interested in, say, the status code.

@benoitc, your thoughts? If any of it makes sense, maybe we can take a stab at it.

**Comment by benoitc** [comment]
if you're only instered in the status why not making an head request then ? Other than that you could also do smth like : `hackney:get(URL, [], <<>>, [with_body])` .

Hopefully it will help, let me know :)

**Comment by swvist** [comment]
Had not used the head request before. That is indeed what we should be using. Missed the with_body part in the docs as well. This solves it for us. Thanks @benoitc :)

@jaseemabid, close the issue?

**Comment by jaseemabid** [comment]
Yes. We must ideally update the docs to make this behaviour clear. It will just deadlock otherwise and it's not trivial to debug.

**Comment by benoitc** [comment]
@jaseemabid that would be nice! Don't forget that the readme is generated from `doc/overview.edoc` though :)

**Comment by benoitc** [comment]
closed via 74fa2a1a412d6b8451442e1180614db1567029eb
